File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/PHashSimple.swift
import Foundation

/// phashSimple: grayscale → Lanczos to (N*F, N*F) → row-wise 1-D DCT → block [0:N, 1:N+1] → mean threshold.
///
/// This matches `imagehash.phash_simple` (Python imagehash 4.3.2) exactly:
///   dct = scipy.fftpack.dct(pixels)          # 1-D DCT applied row-wise (NOT 2-D DCT)
///   dctlowfreq = dct[:hash_size, 1:hash_size+1]   # rows 0..N-1, cols 1..N (skip DC)
///   avg = dctlowfreq.mean()
///   diff = dctlowfreq > avg
///
/// Note: despite the name, this is NOT simply phash with mean instead of median.
/// It uses a 1-D (row-wise) DCT and skips column 0 (DC component).
public func phashSimple(_ image: RGBImage, hashSize: Int, highfreqFactor: Int = 4) throws -> Hash {
	try image.validate()
	guard hashSize >= 2 else { throw ImageHashError.invalidHashSize(hashSize) }
	let imgSize = hashSize * highfreqFactor

	let rgb = ImgRGB.toRGB(image)
	let gray = rgbToGray(rgb.data, width: rgb.width, height: rgb.height)
	let resized = lanczosResize(gray, srcW: rgb.width, srcH: rgb.height, dstW: imgSize, dstH: imgSize)

	// Row-wise 1-D DCT (scipy.fftpack.dct(pixels) applies DCT to each row)
	var dctRows = [[Double]](repeating: [], count: imgSize)
	for y in 0..<imgSize {
		var row = [Double](repeating: 0.0, count: imgSize)
		for x in 0..<imgSize {
			row[x] = Double(resized[y * imgSize + x])
		}
		dctRows[y] = dct1d(row)
	}

	// Extract dct[:hashSize, 1:hashSize+1] — rows 0..<N, cols 1..<N+1 (skip DC column 0)
	var block = [Double](repeating: 0.0, count: hashSize * hashSize)
	var k = 0
	for y in 0..<hashSize {
		for x in 1...hashSize {
			block[k] = dctRows[y][x]
			k += 1
		}
	}

	// Mean threshold with snap-to-threshold tie-break.
	let mean = block.reduce(0.0, +) / Double(block.count)

	// Snap-to-threshold tie-break: deterministic bit 0 on ties.
	// See spec/SPEC.md §"Threshold tie-break".
	let threshold = mean + SNAP_EPS
	var bits: [[Bool]] = []
	for y in 0..<hashSize {
		var row = [Bool](repeating: false, count: hashSize)
		for x in 1...hashSize {
			row[x - 1] = dctRows[y][x] > threshold
		}
		bits.append(row)
	}
	return try Hash(bits: bits)
}

File: hash/java/src/main/java/io/github/wmetcalf/rosettasquint/hash/PHashSimple.java
package io.github.wmetcalf.rosettasquint.hash;

import io.github.wmetcalf.rosettasquint.hash.internal.BufferedImageRgb;
import io.github.wmetcalf.rosettasquint.hash.internal.Dct;
import io.github.wmetcalf.rosettasquint.hash.internal.LanczosFixed;
import io.github.wmetcalf.rosettasquint.hash.internal.PilGrayscale;

import java.awt.image.BufferedImage;

/**
 * phash_simple: row-wise 1-D DCT (not 2-D) → take rows [0,N) cols [1,N+1) → mean threshold.
 *
 * <p>The Python imagehash source (v4.3.2) uses {@code scipy.fftpack.dct(pixels)} with no
 * axis argument, which applies 1-D DCT along the last axis (rows). It then takes
 * {@code dct[:hash_size, 1:hash_size+1]} — skipping the DC column (index 0) —
 * and thresholds against the mean (not the median used by phash).
 *
 * <p>This is distinct from {@code phash} which does a full 2-D DCT (column-then-row)
 * and takes {@code dct[0:N, 0:N]} (including the DC term at [0,0]).
 */
public final class PHashSimple {
    public static ImageHash compute(BufferedImage img, int hashSize) {
        return compute(img, hashSize, 4);
    }

    public static ImageHash compute(BufferedImage img, int hashSize, int highfreqFactor) {
        if (hashSize < 2)
            throw new IllegalArgumentException("hashSize must be >= 2");

        int imgSize = hashSize * highfreqFactor;

        int[][] rgb = BufferedImageRgb.toIntArray(img);
        int h = rgb.length;
        int w = rgb[0].length;
        int[][] gray = new int[h][w];
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                int p = rgb[y][x];
                gray[y][x] = PilGrayscale.toGray((p >> 16) & 0xFF, (p >> 8) & 0xFF, p & 0xFF);
            }
        }

        int[][] resized = LanczosFixed.resize(gray, imgSize, imgSize);

        // Row-wise 1-D DCT (scipy.fftpack.dct(pixels) with default axis=-1)
        // Then take rows [0, hashSize), cols [1, hashSize+1)
        double sum = 0.0;
        double[][] block = new double[hashSize][hashSize];
        for (int y = 0; y < imgSize; y++) {
            // Only process rows needed for the output block
            if (y >= hashSize) break;
            double[] row = new double[imgSize];
            for (int x = 0; x < imgSize; x++) row[x] = resized[y][x];
            double[] dctRow = Dct.dct1d(row);
            // Take cols [1, hashSize+1)
            for (int x = 0; x < hashSize; x++) {
                block[y][x] = dctRow[x + 1];
                sum += dctRow[x + 1];
            }
        }
        double mean = sum / (hashSize * hashSize);

        // Snap-to-threshold tie-break: deterministic bit 0 on ties.
        // See spec/SPEC.md §"Threshold tie-break".
        double threshold = mean + PHash.SNAP_EPS;
        boolean[][] bits = new boolean[hashSize][hashSize];
        for (int y = 0; y < hashSize; y++)
            for (int x = 0; x < hashSize; x++)
                bits[y][x] = block[y][x] > threshold;

        return new ImageHash(bits);
    }

    private PHashSimple() {}
}
